micro_manager/model_adaptivity/model_adaptivity.py

```diff
@@ -0,0 +1,354 @@
+"""
+Class ModelAdaptivity provides methods to change micro simulation resolution on the fly.
+"""
+from typing import Union, Optional
+
+from ..config import Config
+from ..micro_simulation import create_simulation_class
+from micro_manager.tools.logging_wrapper import Logger
+
+import numpy as np
+import importlib
+
+
+class ModelAdaptivity:
+    def __init__(self, configurator: Config, rank: int) -> None:
+        """
+        Class constructor.
+
+        Parameters
+        ----------
+        configurator : object of class Config
+            Object which has getter functions to get parameters defined in the configuration file.
+        rank : int
+            Rank of the MPI communicator.
+        """
+        self._logger = Logger("ModelAdaptivity", "./ModelAdaptivity.log", rank)
+
+        self._model_files = configurator.get_model_adaptivity_file_names()
+        self._switching_func_name = (
+            configurator.get_model_adaptivity_switching_function()
+        )
+
+        self._model_classes = []
+        CLASS_NAME = "MicroSimulation"
+        for model_file in self._model_files:
+            try:
+                model = getattr(
+                    importlib.import_module(model_file, CLASS_NAME),
+                    CLASS_NAME,
+                )
+                self._model_classes.append(create_simulation_class(model))
+            except Exception as e:
+                self._logger.log_info_rank_zero(
+                    f"Failed to load model class with error: {e}"
+                )
+        if (
+            len(self._model_classes) != len(self._model_files)
+            or len(self._model_classes) == 0
+        ):
+            raise RuntimeError("Not all models were loaded. Stopping!")
+
+        FUNC_NAME = "switching_function"
+        self._switching_func = ModelAdaptivity.switching_interface
+        try:
+            self._switching_func = getattr(
+                importlib.import_module(self._switching_func_name, FUNC_NAME), FUNC_NAME
+            )
+        except Exception as e:
+            self._logger.log_info_rank_zero(
+                f"Failed to load switching function with error: {e}"
+            )
+
+        self._converged = False
+
+    @staticmethod
+    def switching_interface(
+        resolution: int,
+        location: np.ndarray,
+        t: float,
+        input: dict,
+        prev_output: Optional[dict],
+    ) -> int:
+        """
+        Switching interface function, use as reference
```

Comment:
The sole purpose of this function is to be copied and modified by the user, correct?